railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,389 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(ctx context.Context, kongConfig sendconfig.Kong, log logr.Logger, kubeConfig *rest.Config) {
+	log.Info("Launching Ingress Status Update Thread.")
+	var wg sync.WaitGroup
+	for {
+		select {
+		case updateDone := <-kongConfig.ConfigDone:
+			log.Info("receive configuration information. Update ingress status %v \n", updateDone)
+			wg.Add(1)
+			go UpdateIngress(ctx, &updateDone, log, kubeConfig, &wg)
+			wg.Wait()
```

Comment:
If we are only starting one goroutine with the `WaitGroup` and then we `Wait()` immediately after starting it it's functionally the same as this:

```suggestion
			UpdateIngress(ctx, &updateDone, log, kubeConfig)
```

I suggest we drop the waitgroup and the additional goroutine altogether as otherwise we're creating a goroutine that the runtime has to schedule which provides us no gains.

Reply:
Communicated this to Shane offline, we need waiting group to avoid goroutine memory leak.